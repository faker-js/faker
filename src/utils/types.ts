/**
 * Type that provides auto-suggestions but also any string.
 *
 * @see https://github.com/microsoft/TypeScript/issues/29729#issuecomment-471566609
 */
export type LiteralUnion<T extends U, U = string> =
  | T
  | (U & { zz_IGNORE_ME?: never });

/**
 * Type that represents a single method/function name of the given type.
 */
export type MethodOf<
  ObjectType,
  Signature extends (...args) => unknown = (...args) => unknown
> = {
  [Key in keyof ObjectType]: ObjectType[Key] extends Signature ? Key : never;
}[keyof ObjectType];

/**
 * Type that represents all method/function names of the given type.
 */
export type MethodsOf<
  ObjectType,
  Signature extends (...args) => unknown = (...args) => unknown
> = ReadonlyArray<MethodOf<ObjectType, Signature>>;
