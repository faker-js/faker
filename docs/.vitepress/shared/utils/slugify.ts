export function slugify(value: string): string {
  // Copied from https://github.com/vuejs/docs/blob/b392b068fb893e3ac6079710fe34decbde7a3be3/src/api/ApiIndex.vue#L50-L65
  return (
    value
      // Replace special characters
      .replace(/[\s~`!@#$%^&*()\-_+=[\]{}|\\;:"'<>,.?/]+/g, '-')
      // Remove continuous separators
      .replace(/\-{2,}/g, '-')
      // Remove prefixing and trailing separators
      .replace(/^\-+|\-+$/g, '')
      // ensure it doesn't start with a number (like #123)
      .replace(/^(\d)/, '_$1')
      // lowercase
      .toLowerCase()
  );
}
