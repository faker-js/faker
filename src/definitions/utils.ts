// https://stackoverflow.com/a/53395649/4573065
export type AllOf<T> = ['Needs to be all of', T];

/**
 * Creates a function that requires all keys of the generic type to be used as parameters.
 * The function itself will return the given parameters.
 */
export function allOf<T>(): <U extends T[]>(
  ...array: U & ([T] extends [U[number]] ? unknown : AllOf<T>[])
) => U & ([T] extends [U[number]] ? unknown : AllOf<T>[]) {
  return (...array) => array;
}
