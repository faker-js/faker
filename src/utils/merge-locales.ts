import type { LocaleDefinition } from '..';

/**
 * Merges the given locales into one locale.
 * The locales are merged in the order they are given.
 * The first locale that provides an entry for a category will be used for that.
 * Mutating the category entries in the returned locale will also mutate the entries in the respective source locale.
 *
 * @param locales The locales to merge.
 * @returns The newly merged locale.
 *
 * @example
 * const de_CH_with_fallbacks = mergeLocales([ de_CH, de, en ]);
 */
export function mergeLocales(locales: LocaleDefinition[]): LocaleDefinition {
  const merged: LocaleDefinition = {} as LocaleDefinition;

  for (const locale of locales) {
    for (const key in locale) {
      if (merged[key] === undefined) {
        if (typeof locale[key] === 'object') {
          merged[key] = { ...locale[key] };
        } else {
          merged[key] = locale[key];
        }
      } else {
        if (typeof locale[key] === 'object') {
          merged[key] = { ...locale[key], ...merged[key] };
        } else {
          // Primitive values cannot be merged
        }
      }
    }
  }

  return merged;
}
