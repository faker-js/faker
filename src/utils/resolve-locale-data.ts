import type { FakerCore } from '../core';
import type { DefinedLocaleDefinition } from '../definitions/definitions';
import { assertLocaleData } from '../internal/locale-proxy';

/**
 * Returns the locale data for the given category and entry.
 *
 * If the category or entry is missing or explicitly absent, an error is thrown.
 *
 * @param core The core to get the locale data from.
 * @param category  The category to get the locale data from.
 * @param entry The entry to get the locale data from.
 *
 * @example
 * resolveLocaleData(fakerCore, 'food', 'fruit'); // ['apple', 'apricot', ...]
 *
 * @since v10.0.0
 */
export function resolveLocaleData<
  const TCategory extends keyof DefinedLocaleDefinition,
  const TEntry extends keyof DefinedLocaleDefinition[TCategory],
>(
  core: FakerCore,
  category: TCategory,
  entry: TEntry
): DefinedLocaleDefinition[TCategory][TEntry];
/**
 * Returns the locale data for the given category and entry.
 *
 * If the category or entry is missing or explicitly absent, an error is thrown.
 *
 * @param core The core to get the locale data from.
 * @param category  The category to get the locale data from.
 * @param entry The entry to get the locale data from.
 *
 * @example
 * resolveLocaleData(fakerCore, 'food', 'fruit'); // ['apple', 'apricot', ...]
 *
 * @since v10.0.0
 */
export function resolveLocaleData(
  core: FakerCore,
  category: string,
  entry: string
): unknown;
/**
 * Returns the locale data for the given category and entry.
 *
 * If the category or entry is missing or explicitly absent, an error is thrown.
 *
 * @param core The core to get the locale data from.
 * @param category  The category to get the locale data from.
 * @param entry The entry to get the locale data from.
 *
 * @example
 * resolveLocaleData(fakerCore, 'food', 'fruit'); // ['apple', 'apricot', ...]
 *
 * @since v10.0.0
 */
export function resolveLocaleData(
  core: FakerCore,
  category: string,
  entry: string
): unknown {
  const value = core.locale[category]?.[entry];
  assertLocaleData(value, category, entry);
  return value;
}
