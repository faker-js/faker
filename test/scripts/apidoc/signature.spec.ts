import { beforeAll, describe, expect, it } from 'vitest';
import {
  analyzeSignature,
  initMarkdownRenderer,
} from '../../../scripts/apidoc/signature';
import { SignatureTest } from './signature.example';
import { loadExampleMethods } from './utils';

describe('signature', () => {
  describe('analyzeSignature()', () => {
    const methods = loadExampleMethods();

    beforeAll(async () => {
      await initMarkdownRenderer();
    });

    it('dummy dependency to rerun the test if the example changes', () => {
      expect(new SignatureTest()).toBeTruthy();
    });

    it('expected and actual methods are equal', () => {
      expect(Object.keys(methods).sort()).toMatchSnapshot();
    });

    it.each(Object.keys(methods).sort())('%s', (name) => {
      const method = methods[name];
      expect(method, `Method ${name} to be defined`).toBeDefined();
      const actual = analyzeSignature(method.signatures[0], null, method.name);

      expect(actual).toMatchSnapshot();
    });
  });
});
