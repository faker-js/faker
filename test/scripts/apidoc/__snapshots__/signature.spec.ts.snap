// Vitest Snapshot v1

exports[`signature > analyzeSignature() > defaultBooleanParamMethod 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with a default parameter.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">defaultBooleanParamMethod</span><span style=\\"color: #A6ACCD\\">(c: boolean </span><span style=\\"color: #89DDFF\\">=</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #FF9CAC\\">true</span><span style=\\"color: #A6ACCD\\">): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "defaultBooleanParamMethod",
  "parameters": [
    {
      "default": "true",
      "description": "<p>The boolean parameter.</p>
",
      "name": "c",
      "type": "boolean",
    },
  ],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "Default Boolean Param Method",
}
`;

exports[`signature > analyzeSignature() > expected and actual methods are equal 1`] = `
[
  "defaultBooleanParamMethod",
  "functionParamMethod",
  "literalUnionParamMethod",
  "methodWithDeprecated",
  "methodWithExample",
  "methodWithMultipleSeeMarkers",
  "methodWithSinceMarker",
  "multiParamMethod",
  "noParamMethod",
  "optionalStringParamMethod",
  "optionsInlineParamMethodWithDefaults",
  "optionsInterfaceParamMethodWithDefaults",
  "optionsParamMethod",
  "optionsTypeParamMethodWithDefaults",
  "requiredNumberParamMethod",
  "stringUnionParamMethod",
]
`;

exports[`signature > analyzeSignature() > functionParamMethod 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with a function parameters.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">functionParamMethod</span><span style=\\"color: #A6ACCD\\">(fn: </span><span style=\\"color: #89DDFF\\">(</span><span style=\\"color: #A6ACCD\\">a</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #FFCB6B\\">string</span><span style=\\"color: #89DDFF\\">)</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #C792EA\\">=&gt;</span><span style=\\"color: #A6ACCD\\"> number): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "functionParamMethod",
  "parameters": [
    {
      "default": undefined,
      "description": "<p>The function parameter.</p>
",
      "name": "fn",
      "type": "(a: string) => number",
    },
  ],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "Function Param Method",
}
`;

exports[`signature > analyzeSignature() > literalUnionParamMethod 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with LiteralUnion.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">literalUnionParamMethod</span><span style=\\"color: #A6ACCD\\">(value: </span><span style=\\"color: #89DDFF\\">&#39;</span><span style=\\"color: #C3E88D\\">a</span><span style=\\"color: #89DDFF\\">&#39;</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">|</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">&#39;</span><span style=\\"color: #C3E88D\\">b</span><span style=\\"color: #89DDFF\\">&#39;</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">|</span><span style=\\"color: #A6ACCD\\"> string): string</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "literalUnionParamMethod",
  "parameters": [
    {
      "default": undefined,
      "description": "<p><code>'a'</code> or <code>'b'</code>.</p>
",
      "name": "value",
      "type": "'a' | 'b' | string",
    },
  ],
  "returns": "string",
  "seeAlsos": [],
  "since": "",
  "title": "Literal Union Param Method",
}
`;

exports[`signature > analyzeSignature() > methodWithDeprecated 1`] = `
{
  "deprecated": true,
  "description": "<p>Test with deprecated and see marker.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">methodWithDeprecated</span><span style=\\"color: #A6ACCD\\">(): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "methodWithDeprecated",
  "parameters": [],
  "returns": "number",
  "seeAlsos": [
    "test.apidoc.methodWithExample()",
  ],
  "since": "",
  "title": "Method With Deprecated",
}
`;

exports[`signature > analyzeSignature() > methodWithExample 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with example marker.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">methodWithExample</span><span style=\\"color: #A6ACCD\\">(): number</span></span>
<span class=\\"line\\"><span style=\\"color: #A6ACCD\\">test</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #A6ACCD\\">apidoc</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">methodWithExample</span><span style=\\"color: #A6ACCD\\">() </span><span style=\\"color: #676E95\\">// 0</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "methodWithExample",
  "parameters": [],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "Method With Example",
}
`;

exports[`signature > analyzeSignature() > methodWithMultipleSeeMarkers 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with multiple see markers.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">methodWithMultipleSeeMarkers</span><span style=\\"color: #A6ACCD\\">(): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "methodWithMultipleSeeMarkers",
  "parameters": [],
  "returns": "number",
  "seeAlsos": [
    "test.apidoc.methodWithExample()",
    "test.apidoc.methodWithDeprecated()",
  ],
  "since": "",
  "title": "Method With Multiple See Markers",
}
`;

exports[`signature > analyzeSignature() > methodWithSinceMarker 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with since marker.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">methodWithSinceMarker</span><span style=\\"color: #A6ACCD\\">(): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "methodWithSinceMarker",
  "parameters": [],
  "returns": "number",
  "seeAlsos": [],
  "since": "1.0.0",
  "title": "Method With Since Marker",
}
`;

exports[`signature > analyzeSignature() > multiParamMethod 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with multiple parameters.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">multiParamMethod</span><span style=\\"color: #A6ACCD\\">(a: number</span><span style=\\"color: #89DDFF\\">,</span><span style=\\"color: #A6ACCD\\"> b</span><span style=\\"color: #89DDFF\\">?:</span><span style=\\"color: #A6ACCD\\"> string</span><span style=\\"color: #89DDFF\\">,</span><span style=\\"color: #A6ACCD\\"> c: boolean </span><span style=\\"color: #89DDFF\\">=</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #FF9CAC\\">true</span><span style=\\"color: #A6ACCD\\">): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "multiParamMethod",
  "parameters": [
    {
      "default": undefined,
      "description": "<p>The number parameter.</p>
",
      "name": "a",
      "type": "number",
    },
    {
      "default": undefined,
      "description": "<p>The string parameter.</p>
",
      "name": "b?",
      "type": "string",
    },
    {
      "default": "true",
      "description": "<p>The boolean parameter.</p>
",
      "name": "c",
      "type": "boolean",
    },
  ],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "Multi Param Method",
}
`;

exports[`signature > analyzeSignature() > noParamMethod 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with no parameters.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">noParamMethod</span><span style=\\"color: #A6ACCD\\">(): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "noParamMethod",
  "parameters": [],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "No Param Method",
}
`;

exports[`signature > analyzeSignature() > optionalStringParamMethod 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with an optional parameter.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">optionalStringParamMethod</span><span style=\\"color: #A6ACCD\\">(b</span><span style=\\"color: #89DDFF\\">?:</span><span style=\\"color: #A6ACCD\\"> string): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "optionalStringParamMethod",
  "parameters": [
    {
      "default": undefined,
      "description": "<p>The string parameter.</p>
",
      "name": "b?",
      "type": "string",
    },
  ],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "Optional String Param Method",
}
`;

exports[`signature > analyzeSignature() > optionsInlineParamMethodWithDefaults 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with a function parameters (inline types) with defaults.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">optionsInlineParamMethodWithDefaults</span><span style=\\"color: #A6ACCD\\">(a: </span><span style=\\"color: #89DDFF\\">{</span></span>
<span class=\\"line\\"><span style=\\"color: #A6ACCD\\">  </span><span style=\\"color: #F07178\\">value</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> number</span></span>
<span class=\\"line\\"><span style=\\"color: #89DDFF\\">}</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">=</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">{</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F07178\\">value</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F78C6C\\">1</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">},</span><span style=\\"color: #A6ACCD\\"> b: </span><span style=\\"color: #89DDFF\\">{</span></span>
<span class=\\"line\\"><span style=\\"color: #A6ACCD\\">  </span><span style=\\"color: #F07178\\">value</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> number</span></span>
<span class=\\"line\\"><span style=\\"color: #89DDFF\\">}</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">=</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">{</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F07178\\">value</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F78C6C\\">1</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">},</span><span style=\\"color: #A6ACCD\\"> c: </span><span style=\\"color: #89DDFF\\">{</span></span>
<span class=\\"line\\"><span style=\\"color: #A6ACCD\\">  </span><span style=\\"color: #F07178\\">value</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> number</span></span>
<span class=\\"line\\"><span style=\\"color: #89DDFF\\">}</span><span style=\\"color: #A6ACCD\\">): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "optionsInlineParamMethodWithDefaults",
  "parameters": [
    {
      "default": "{ value: 1 }",
      "description": "<p>Parameter with signature default.
It also has a more complex description.</p>
",
      "name": "a",
      "type": "{ ... }",
    },
    {
      "default": undefined,
      "description": "<p>The number parameter.</p>
",
      "name": "a.value?",
      "type": "number",
    },
    {
      "default": "{ value: 1 }",
      "description": "<p>Parameter with jsdocs default.</p>
<p>It also has a more complex description.</p>
",
      "name": "b",
      "type": "{ ... }",
    },
    {
      "default": undefined,
      "description": "<p>The number parameter.</p>
",
      "name": "b.value?",
      "type": "number",
    },
    {
      "default": undefined,
      "description": "<p>Parameter with inner jsdocs default.</p>
",
      "name": "c",
      "type": "{ ... }",
    },
    {
      "default": "2",
      "description": "<p>The number parameter. It also has a more complex description.</p>
",
      "name": "c.value?",
      "type": "number",
    },
  ],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "Options Inline Param Method With Defaults",
}
`;

exports[`signature > analyzeSignature() > optionsInterfaceParamMethodWithDefaults 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with a function parameters with defaults.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">optionsInterfaceParamMethodWithDefaults</span><span style=\\"color: #A6ACCD\\">(a: ParameterOptionsInterfaceA </span><span style=\\"color: #89DDFF\\">=</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">{</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F07178\\">value</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F78C6C\\">1</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">},</span><span style=\\"color: #A6ACCD\\"> b: ParameterOptionsInterfaceB </span><span style=\\"color: #89DDFF\\">=</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">{</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F07178\\">value</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F78C6C\\">1</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">},</span><span style=\\"color: #A6ACCD\\"> c: ParameterOptionsInterfaceC): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "optionsInterfaceParamMethodWithDefaults",
  "parameters": [
    {
      "default": "{ value: 1 }",
      "description": "<p>Parameter with signature default.</p>
",
      "name": "a",
      "type": "ParameterOptionsInterfaceA",
    },
    {
      "default": "{ value: 1 }",
      "description": "<p>Parameter with jsdocs default.</p>
",
      "name": "b",
      "type": "ParameterOptionsInterfaceB",
    },
    {
      "default": undefined,
      "description": "<p>Parameter with inner jsdocs default.</p>
",
      "name": "c",
      "type": "ParameterOptionsInterfaceC",
    },
  ],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "Options Interface Param Method With Defaults",
}
`;

exports[`signature > analyzeSignature() > optionsParamMethod 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with a function parameters.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">optionsParamMethod</span><span style=\\"color: #A6ACCD\\">(options: </span><span style=\\"color: #89DDFF\\">{</span></span>
<span class=\\"line\\"><span style=\\"color: #A6ACCD\\">  </span><span style=\\"color: #F07178\\">a</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> number</span><span style=\\"color: #89DDFF\\">,</span></span>
<span class=\\"line\\"><span style=\\"color: #A6ACCD\\">  </span><span style=\\"color: #F07178\\">b</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> string</span><span style=\\"color: #89DDFF\\">,</span></span>
<span class=\\"line\\"><span style=\\"color: #A6ACCD\\">  </span><span style=\\"color: #F07178\\">c</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> boolean</span><span style=\\"color: #89DDFF\\">,</span></span>
<span class=\\"line\\"><span style=\\"color: #A6ACCD\\">  </span><span style=\\"color: #82AAFF\\">d</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">()</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #C792EA\\">=&gt;</span><span style=\\"color: #A6ACCD\\"> string</span></span>
<span class=\\"line\\"><span style=\\"color: #89DDFF\\">}</span><span style=\\"color: #A6ACCD\\">): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "optionsParamMethod",
  "parameters": [
    {
      "default": undefined,
      "description": "<p>The function parameter.</p>
",
      "name": "options",
      "type": "{ ... }",
    },
    {
      "default": undefined,
      "description": "<p>The number parameter.</p>
",
      "name": "options.a",
      "type": "number",
    },
    {
      "default": undefined,
      "description": "<p>The string parameter.</p>
",
      "name": "options.b?",
      "type": "string",
    },
    {
      "default": undefined,
      "description": "<p>The boolean parameter.</p>
",
      "name": "options.c",
      "type": "boolean",
    },
    {
      "default": undefined,
      "description": "<p>The method parameter.</p>
",
      "name": "options.d",
      "type": "() => string",
    },
  ],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "Options Param Method",
}
`;

exports[`signature > analyzeSignature() > optionsTypeParamMethodWithDefaults 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with a function parameters with defaults.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">optionsTypeParamMethodWithDefaults</span><span style=\\"color: #A6ACCD\\">(a: ParameterOptionsTypeA </span><span style=\\"color: #89DDFF\\">=</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">{</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F07178\\">value</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F78C6C\\">1</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">},</span><span style=\\"color: #A6ACCD\\"> b: ParameterOptionsTypeB </span><span style=\\"color: #89DDFF\\">=</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">{</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F07178\\">value</span><span style=\\"color: #89DDFF\\">:</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #F78C6C\\">1</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">},</span><span style=\\"color: #A6ACCD\\"> c: ParameterOptionsTypeC): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "optionsTypeParamMethodWithDefaults",
  "parameters": [
    {
      "default": "{ value: 1 }",
      "description": "<p>Parameter with signature default.</p>
",
      "name": "a",
      "type": "ParameterOptionsTypeA",
    },
    {
      "default": "{ value: 1 }",
      "description": "<p>Parameter with jsdocs default.</p>
",
      "name": "b",
      "type": "ParameterOptionsTypeB",
    },
    {
      "default": undefined,
      "description": "<p>Parameter with inner jsdocs default.</p>
",
      "name": "c",
      "type": "ParameterOptionsTypeC",
    },
  ],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "Options Type Param Method With Defaults",
}
`;

exports[`signature > analyzeSignature() > requiredNumberParamMethod 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with a required parameter.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">requiredNumberParamMethod</span><span style=\\"color: #A6ACCD\\">(a: number): number</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "requiredNumberParamMethod",
  "parameters": [
    {
      "default": undefined,
      "description": "<p>The number parameter.</p>
",
      "name": "a",
      "type": "number",
    },
  ],
  "returns": "number",
  "seeAlsos": [],
  "since": "",
  "title": "Required Number Param Method",
}
`;

exports[`signature > analyzeSignature() > stringUnionParamMethod 1`] = `
{
  "deprecated": false,
  "description": "<p>Test with string union.</p>
",
  "examples": "<div class=\\"language-ts\\"><button title=\\"Copy Code\\" class=\\"copy\\"></button><span class=\\"lang\\">ts</span><pre v-pre class=\\"shiki\\" ><code><span class=\\"line\\"><span style=\\"color: #A6ACCD\\">faker</span><span style=\\"color: #89DDFF\\">.</span><span style=\\"color: #82AAFF\\">stringUnionParamMethod</span><span style=\\"color: #A6ACCD\\">(value: </span><span style=\\"color: #89DDFF\\">&#39;</span><span style=\\"color: #C3E88D\\">a</span><span style=\\"color: #89DDFF\\">&#39;</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">|</span><span style=\\"color: #A6ACCD\\"> </span><span style=\\"color: #89DDFF\\">&#39;</span><span style=\\"color: #C3E88D\\">b</span><span style=\\"color: #89DDFF\\">&#39;</span><span style=\\"color: #A6ACCD\\">): string</span></span>
<span class=\\"line\\"></span></code></pre>
</div>",
  "name": "stringUnionParamMethod",
  "parameters": [
    {
      "default": undefined,
      "description": "<p><code>'a'</code> or <code>'b'</code>.</p>
",
      "name": "value",
      "type": "'a' | 'b'",
    },
  ],
  "returns": "string",
  "seeAlsos": [],
  "since": "",
  "title": "String Union Param Method",
}
`;
